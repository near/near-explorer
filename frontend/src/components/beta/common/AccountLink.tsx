import * as React from "react";
import Link from "../../utils/Link";
import { shortenString } from "../../../libraries/formatting";
import { styled } from "../../../libraries/styles";

const AccountLinkWrapper = styled("a", {
  whiteSpace: "nowrap",
});

export interface Props {
  accountId: string;
}

const AccountLink: React.FC<Props> = React.memo(({ accountId }) => {
  return (
    <Link href={`/beta/accounts/${accountId}`} passHref>
      <AccountLinkWrapper>{shortenString(accountId)}</AccountLinkWrapper>
    </Link>
  );
});

export default AccountLink;
