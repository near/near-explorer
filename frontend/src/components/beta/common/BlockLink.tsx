import * as React from "react";
import Link from "../../utils/Link";
import { styled } from "../../../libraries/styles";

export interface Props {
  blockHash: string;
  blockHeight: number;
}

const LinkWrapper = styled("a", {
  whiteSpace: "nowrap",
  cursor: "pointer",
});

const BlockLink: React.FC<Props> = React.memo(({ blockHash, blockHeight }) => (
  <Link href="/blocks/[hash]" as={`/blocks/${blockHash}`}>
    <LinkWrapper>{`#${blockHeight}`}</LinkWrapper>
  </Link>
));

export default BlockLink;
