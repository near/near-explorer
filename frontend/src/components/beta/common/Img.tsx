import * as React from "react";
import Image from "next/image";
import { styled } from "../../../libraries/styles";
import { rgbDataURL } from "../../../libraries/rgbplaceholder";

const FallbackImage = styled("img", {
  width: "100%",
});

type Props = {
  src: string | null;
};
const Img: React.FC<Props> = React.memo(({ src }) => {
  const [mediaUrl, setMediaUrl] = React.useState(src ?? null);
  const handleMediaUrl = React.useCallback(
    (url) => () => setMediaUrl(url),
    [setMediaUrl]
  );

  if (!mediaUrl) {
    return <FallbackImage src="/static/images/failed_to_load.svg" />;
  }

  return (
    <Image
      src={mediaUrl}
      layout="fill"
      placeholder="blur"
      blurDataURL={rgbDataURL(220, 220, 220)}
      onError={handleMediaUrl("/static/images/failed_to_load.svg")}
      onLoadingComplete={handleMediaUrl(src)}
    />
  );
});

export default Img;
