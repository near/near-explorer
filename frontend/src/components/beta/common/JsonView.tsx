import dynamic from "next/dynamic";
import * as React from "react";
// https://github.com/mac-s-g/react-json-view/issues/296#issuecomment-803497117
const DynamicReactJson = dynamic(import("react-json-view"), { ssr: false });

type Props = {
  args: object;
};

const JsonView: React.FC<Props> = React.memo(({ args }) => (
  <DynamicReactJson
    src={args}
    name={null}
    iconStyle="triangle"
    displayObjectSize={false}
    displayDataTypes={false}
    style={{
      fontSize: "14px",
      padding: "10px 0",
    }}
  />
));

export default JsonView;
