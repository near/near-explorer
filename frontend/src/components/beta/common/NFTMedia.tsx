import * as React from "react";
import { Img } from "react-image";
import { styled } from "../../../libraries/styles";
import PaginationSpinner from "../../utils/PaginationSpinner";

const Image = styled(Img, {
  width: "100%",
});

const FallbackImage = styled(Image);

type Props = {
  src: string | null;
};
const NFTMedia: React.FC<Props> = React.memo(({ src }) => {
  const [mediaUrl, setMediaUrl] = React.useState(src ?? null);
  const handleMediaUrl = React.useCallback(
    () => setMediaUrl("/static/images/failed_to_load.svg"),
    [setMediaUrl]
  );
  const [poster, setPoster] = React.useState("");
  const handlePoster = React.useCallback(
    () => setPoster("/static/images/failed_to_load.svg"),
    [setPoster]
  );

  if (!mediaUrl) {
    return <FallbackImage src="/static/images/failed_to_load.svg" />;
  }
  // copied from wallet https://github.com/near/near-wallet/blob/master/packages/frontend/src/components/nft/NFTMedia.js
  const [isVideo, mimeType] = React.useMemo(() => {
    let mimeType;
    // check mediaUrl string for .webm or .mp4 endings (case-insensitive)
    if (mediaUrl.match(/\.webm$/i)) mimeType = "webm";
    else if (mediaUrl.match(/\.mp4$/i)) mimeType = "mp4";
    // if there is a mediaUrl and a truthy mimeType (webm or mp4), we have a video
    const isVideo = !!mediaUrl && mimeType;
    return [isVideo, mimeType];
  }, [mediaUrl]);

  return (
    <>
      {isVideo ? (
        <video muted={true} loop controls autoPlay={false} poster={poster}>
          <source
            src={mediaUrl}
            type={`video/${mimeType}`}
            onError={handlePoster}
          />
        </video>
      ) : (
        <Image
          alt="NFT"
          src={mediaUrl}
          loader={<PaginationSpinner />}
          onError={handleMediaUrl}
        />
      )}
    </>
  );
});

export default NFTMedia;
