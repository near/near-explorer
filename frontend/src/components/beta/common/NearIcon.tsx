import * as React from "react";
import { styled } from "../../../libraries/styles";

const Wrapper = styled("span", { display: "inline-flex" });

const NearIcon: React.FC = React.memo(() => {
  return (
    <Wrapper>
      <svg width=".8em" height=".8em" fill="none" viewBox="0 0 15 15">
        <path
          d="m12.032 0.762-3.135 4.655c-0.05 0.067-0.074 0.151-0.065 0.235 8e-3 0.083 0.048 0.161 0.11147 0.21644 0.06331 0.05554 0.14528 0.08503 0.22944 0.08254 0.08415-0.00249 0.16424-0.03678 0.22416-0.09596l3.0846-2.6667c0.0178-0.01632 0.0399-0.02706 0.0638-0.03088 0.0238-0.00382 0.0482-5.6e-4 0.0702 0.00937 0.022 0.00994 0.0406 0.02612 0.0535 0.04654 0.0129 0.02041 0.0195 0.04416 0.019 0.0683v8.3816c-2e-4 0.0255-0.0083 0.0503-0.0229 0.071-0.0147 0.0208-0.0354 0.0366-0.0593 0.0453s-0.0499 0.0099-0.0745 0.0034-0.0466-0.0203-0.0631-0.0397l-9.3272-11.167c-0.1489-0.17927-0.33519-0.32378-0.5458-0.4234-0.21061-0.099616-0.44044-0.15193-0.67339-0.15327h-0.32478c-0.42362 0-0.8299 0.16839-1.1294 0.46814s-0.46783 0.70629-0.46783 1.1302v11.803c0 0.4239 0.16828 0.8304 0.46783 1.1302 0.29955 0.2997 0.70582 0.4681 1.1294 0.4681 0.27295 0 0.54134-0.0701 0.77956-0.2034s0.43832-0.3255 0.5812-0.5583l3.1346-4.655c0.05079-0.06718 0.07427-0.15112 0.06573-0.23494-0.00855-0.08381-0.04849-0.16127-0.1118-0.21681s-0.14529-0.08502-0.22944-0.08253-0.16424 0.03677-0.22416 0.09595l-3.0846 2.6667c-0.01779 0.0163-0.03997 0.027-0.06379 0.0309-0.02383 0.0038-0.04825 5e-4 -0.07024-0.0094-0.02199-0.01-0.04059-0.0261-0.05347-0.0466-0.01289-0.0204-0.01951-0.0441-0.01903-0.0683v-8.3733c2.5e-4 -0.02544 0.00826-0.05021 0.02295-0.07098s0.03537-0.03656 0.05927-0.04526c0.0239-0.00869 0.04988-0.00988 0.07448-0.00341 0.02459 0.00648 0.04663 0.02031 0.06315 0.03965l9.3272 11.167c0.15 0.1776 0.337 0.3204 0.5478 0.4182 0.2109 0.0979 0.4406 0.1486 0.673 0.1484h0.3331c0.2098 0 0.4175-0.0413 0.6113-0.1216 0.1938-0.0804 0.3699-0.1981 0.5182-0.3465s0.2659-0.3246 0.3462-0.5185c0.0803-0.194 0.1216-0.4018 0.1216-0.6117v-11.802c0-0.21074-0.0416-0.4194-0.1225-0.61397-0.0809-0.19457-0.1995-0.37122-0.3489-0.51977-0.1493-0.14855-0.3266-0.26607-0.5215-0.34581-0.195-0.079733-0.4038-0.1201-0.6144-0.11878-0.2729 3.2768e-5 -0.5413 0.07006-0.7795 0.2034-0.2382 0.13334-0.4383 0.32554-0.5812 0.55826z"
          fill="currentColor"
        />
      </svg>
    </Wrapper>
  );
});

export default NearIcon;
