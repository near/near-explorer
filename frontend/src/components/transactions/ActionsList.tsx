import * as T from "../../libraries/explorer-wamp/transactions";

import ActionRow, { ViewMode } from "./ActionRow";

export interface Props {
  actions: T.Action[];
  transaction: T.Transaction;
  viewMode?: ViewMode;
}

export default ({ actions, transaction, viewMode }: Props) => (
  <>
    {actions.map((action, actionIndex) => (
      <ActionRow
        key={transaction.hash + actionIndex}
        action={action}
        transaction={transaction}
        viewMode={viewMode}
      />
    ))}
  </>
);
