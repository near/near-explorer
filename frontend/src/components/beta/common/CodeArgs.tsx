import * as React from "react";
import { hexy } from "hexy";

import { styled } from "../../../libraries/styles";

import CodePreview from "../../utils/CodePreview";
import JsonView from "./JsonView";

const HexArgs = styled("div", {
  padding: "10px 0",

  "& > div": {
    background: "#f8f8f8",
    borderRadius: 4,
    color: "#3f4246",
    padding: 20,
    fontSize: 14,
    fontWeight: 500,

    "& textarea, pre": {
      background: "inherit",
      color: "inherit",
      fontFamily: "inherit",
      fontSize: "inherit",
      border: "none",
      padding: 0,
    },
  },
});

const CodeArgs: React.FC<{ args: string }> = React.memo(({ args }) => {
  const decodedArgs = Buffer.from(args, "base64");

  let prettyArgs: object | string;
  try {
    const parsedJSONArgs = JSON.parse(decodedArgs.toString());
    prettyArgs =
      typeof parsedJSONArgs === "boolean"
        ? JSON.stringify(parsedJSONArgs)
        : parsedJSONArgs;
  } catch {
    prettyArgs = hexy(decodedArgs, { format: "twos" });
  }
  return typeof prettyArgs === "object" ? (
    <JsonView args={prettyArgs} />
  ) : (
    <HexArgs>
      <CodePreview collapseHeight={200} maxHeight={600} value={prettyArgs} />
    </HexArgs>
  );
});

export default CodeArgs;
