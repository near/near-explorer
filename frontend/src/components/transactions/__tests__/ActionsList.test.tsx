import renderer from "react-test-renderer";

import ActionsList from "../ActionsList";

import { TRANSACTIONS } from "./common";

describe("<ActionsList />", () => {
  it("renders sparsely by default", () => {
    expect(
      renderer.create(
        <ActionsList
          transaction={TRANSACTIONS[0]}
          actions={TRANSACTIONS[0].actions}
        />
      )
    ).toMatchSnapshot();
  });

  it("renders compact", () => {
    expect(
      renderer.create(
        <ActionsList
          viewMode="compact"
          transaction={TRANSACTIONS[0]}
          actions={TRANSACTIONS[0].actions}
        />
      )
    ).toMatchSnapshot();
  });

  it("renders reversed", () => {
    expect(
      JSON.stringify(
        renderer.create(
          <ActionsList
            transaction={TRANSACTIONS[0]}
            actions={TRANSACTIONS[0].actions}
            reversed
          />
        )
      )
    ).toEqual(
      JSON.stringify(
        renderer.create(
          <ActionsList
            transaction={TRANSACTIONS[0]}
            actions={[...TRANSACTIONS[0].actions].reverse()}
          />
        )
      )
    );
  });

  it("renders functioncall by default", () => {
    expect(
      renderer.create(
        <ActionsList
          transaction={TRANSACTIONS[1]}
          actions={TRANSACTIONS[1].actions}
          showDetails
        />
      )
    ).toMatchSnapshot();
  });
});
