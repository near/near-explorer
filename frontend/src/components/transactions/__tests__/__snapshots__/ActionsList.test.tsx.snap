// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionsList /> renders compact 1`] = `null`;

exports[`<ActionsList /> renders sparsely by default 1`] = `null`;
